controllers/report_controller.go

```diff
@@ -107,3 +154,217 @@ func requeueReportsHandler(c client.Client, log logr.Logger) handler.EventHandle
 		return res
 	})
 }
+
+type reportSchedule interface {
+	// Return the next activation time, later than the given time.
+	// Next is invoked initially, and then each time the job runs..
+	Next(time.Time) time.Time
+}
+
+type reportPeriod struct {
+	periodEnd   time.Time
+	periodStart time.Time
+}
+
+func getReportPeriod(now time.Time, logger logr.Logger, report *curatorv1alpha1.Report) (*reportPeriod, error) {
+	var reportPeriod *reportPeriod
+
+	// check if the report's schedule spec is set
+	if report.Spec.Schedule != nil {
+		reportSchedule, err := getSchedule(report.Spec.Schedule)
+		if err != nil {
+			return nil, err
+		}
+
+		if report.Status.LastReportTime != nil {
+			reportPeriod = getNextReportPeriod(reportSchedule, report.Status.LastReportTime.Time)
+		} else {
+			if report.Spec.ReportingStart != nil {
+				//logger.Info("no last report time for report, using spec.reportingStart %s as starting point", report.Spec.ReportingStart.Time)
+				reportPeriod = getNextReportPeriod(reportSchedule, report.Spec.ReportingStart.Time)
+			} else if report.Status.NextReportTime != nil {
+				//logger.Info("no last report time for report, using status.nextReportTime %s as starting point", report.Status.NextReportTime.Time)
+				reportPeriod = getNextReportPeriod(reportSchedule, report.Status.NextReportTime.Time)
+			} else {
+				// the current period, [now, nextScheduledTime]
+				currentPeriod := getNextReportPeriod(reportSchedule, now)
+				// the next full report period from [nextScheduledTime, nextScheduledTime+1]
+				reportPeriod = getNextReportPeriod(reportSchedule, currentPeriod.periodEnd)
+				report.Status.NextReportTime = &metav1.Time{Time: reportPeriod.periodStart}
+			}
+		}
+	} else {
+		var err error
+		// if there's the Spec.Schedule field is unset, then the report must be a run-once report
+		reportPeriod, err = getRunOnceReportPeriod(report)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
nit: can we move this logic into it's own method for determining the report period? this should reduce the complexity of this getReportPeriod function, and remove the if/else conditional heavy nesting that decreases readability quite a bit.